internal/discoverer/zookeeper.go

```diff
@@ -0,0 +1,297 @@
+	if !ok {
+		return errors.New("Zookeeper service: " + serviceName + " undefined")
+	}
+
+	zd.sendMessage(zkService.(*ZookeeperService), make([]Node, 0))
+	zd.removeWatchService(zkService.(*ZookeeperService))
+
+	return nil
+}
+
+// sendMessage send message notify
+func (zd *ZookeeperDiscoverer) sendMessage(zkService *ZookeeperService, nodes []Node) {
+	messageService := &Service{
+		name:     zkService.Name,
+		nodes:    nodes,
+		entities: make(map[string]struct{}),
+	}
+
+	for _, entry := range zkService.BindEntities {
+		messageService.entities[entry] = struct{}{}
+	}
+
+	msg, err := messageService.NewNotifyMessage()
+	if err != nil {
+		log.Errorf("Zookeeper send message fail, err: %s", err)
+		return
+	}
+
+	zd.msgCh <- msg
+}
+
+// NewZookeeperDiscoverer generate zookeeper discoverer instance
+func NewZookeeperDiscoverer(disConfig interface{}) (Discoverer, error) {
+	config := disConfig.(*conf.Zookeeper)
+
+	conn, _, err := zk.Connect(config.Hosts, time.Second*time.Duration(config.Timeout))
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	discoverer := ZookeeperDiscoverer{
+		msgCh:             make(chan *comm.Message, 10),
+		zkConfig:          config,
+		zkConn:            conn,
+		zkWatchServices:   sync.Map{},
+		zkUnWatchServices: sync.Map{},
+		zkUnWatchContext:  ctx,
+		zkUnWatchCancel:   cancel,
+	}
+
+	err = discoverer.initZookeeperRoot()
+	if err != nil {
+		return nil, err
+	}
+
+	go discoverer.watchServicePrefix()
+
+	return &discoverer, nil
+}
+
+// initZookeeperRoot generate zookeeper root path
+func (zd *ZookeeperDiscoverer) initZookeeperRoot() error {
+	ok, _, err := zd.zkConn.Exists(zd.zkConfig.Prefix)
+	if err != nil {
+		return err
+	}
+
+	if !ok {
+		_, err = zd.zkConn.Create(zd.zkConfig.Prefix, []byte(""), 0, zk.WorldACL(zk.PermAll))
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// newZookeeperClient grantate zookeeper client
+func (zd *ZookeeperDiscoverer) newZookeeperClient(serviceName string) (*ZookeeperService, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	watchPath := zd.zkConfig.Prefix + "/" + serviceName
+	service := &ZookeeperService{
+		Name:         serviceName,
+		BindEntities: make([]string, 0),
+		WatchPath:    watchPath,
+		WatchContext: ctx,
+		WatchCancel:  cancel,
+	}
+
+	return service, nil
+}
+
+// watchServicePrefix watch service prefix change, update unwatch service
+func (zd *ZookeeperDiscoverer) watchServicePrefix() {
+	for {
+		_, _, event, err := zd.zkConn.ChildrenW(zd.zkConfig.Prefix)
+		if err != nil {
+			log.Errorf("watch service prefix: %s fail, err: %s", zd.zkConfig.Prefix, err)
+			return
+		}
+
+		select {
+		case <-zd.zkUnWatchContext.Done():
+			return
+		case <-event:
+			var serviceNames []string
+			serviceNames, _, err = zd.zkConn.Children(zd.zkConfig.Prefix)
+			if err != nil {
+				log.Errorf("fetch service prefix: %s fail, err: %s", zd.zkConfig.Prefix, err)
+				return
```

Comment:
We should use `continue` here?

Reply:
Yes, using continue will not cause the watch of the prefix to fail. updated